### PR TITLE
fix: propagate contextual tuples to ListObjects with concurrent checks

### PR DESCRIPTION
**Pull request by jon-whit**
## Description
This fixes an issue that leads to ListObjects reporting different results than Check if the unoptimized version of ListObjects is delegated to.

## References
https://github.com/openfga/openfga/issues/557

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [ ] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: pkg/server/test/list_objects.go [hunk 1]
```diff
@@ -51,6 +52,63 @@ func newListObjectsRequest(store, objectType, relation, user, modelID string, co
 func ListObjectsTest(t *testing.T, ds storage.OpenFGADatastore) {
 	ctx := context.Background()
 
+	t.Run("list_objects_with_concurrent_checks_and_contextual_tuples", func(t *testing.T) {
```
Comment (jon-whit) [inline pkg/server/test/list_objects.go:55]: Added this here as a means to ensure we don't fail this issue again in the future. I'm working on a separate PR to add more functional tests for ListObjects like we've done with Check 👍 

In the future this test will move into that other test suite once it is done.

File: pkg/server/test/list_objects.go [hunk 2]
```diff
@@ -141,21 +201,6 @@ func ListObjectsTest(t *testing.T, ds storage.OpenFGADatastore) {
 				expectedResult: []string{},
 				expectedError:  nil,
 			},
-			{
-				name: "returns_error_if_duplicate_contextual_tuples",
```
Comment (jon-whit) [inline pkg/server/test/list_objects.go:145]: Removed this test and functionality associated with it. If the developer wants to provide a duplicate contextual tuple, then so be it. I don't see a reason why we should return an error if they provide a duplicate.

If they want to provide duplicates, then that should be fine so long as they still meet the maximum contextual tuple request limits.

File: pkg/server/test/list_objects.go [hunk 3]
```diff
@@ -51,6 +52,63 @@ func newListObjectsRequest(store, objectType, relation, user, modelID string, co
...
+			Id:              ulid.Make().String(),
+			SchemaVersion:   typesystem.SchemaVersion1_1,
+			TypeDefinitions: typedefs,
+		}
+		err := ds.WriteAuthorizationModel(ctx, store, model)
+		require.NoError(t, err)
+
+		typesys := typesystem.New(model)
+
+		listObjectsQuery := &commands.ListObjectsQuery{
```
Comment (jon-whit) [inline pkg/server/test/list_objects.go:78]: Note to self - we have this in a lot of places. We should consolidate to

```
func NewListObjectsQuery(...) {
    // set defaults if none are overriden (e.g. `NoopLogger`, `defaultListObjectsDeadline`, etc..)
}
```

this will cut down on boilerplate in a lot of places.

File: pkg/server/test/list_objects.go [hunk 4]
```diff
@@ -51,6 +52,63 @@ func newListObjectsRequest(store, objectType, relation, user, modelID string, co
...
+			User:                 "user:jon",
+			ContextualTuples: &openfgapb.ContextualTupleKeys{
+				TupleKeys: []*openfgapb.TupleKey{
+					tuple.NewTupleKey("doc:1", "reader", "user:jon"),
+					tuple.NewTupleKey("doc:2", "reader", "user:jon"),
+				},
+			},
+		}
+
+		ctx = typesystem.ContextWithTypesystem(ctx, typesys)
```
Comment (adriantam) [inline pkg/server/test/list_objects.go:101]: a bit sad that we have to do this in the test instead of calling server's code to ensure this is setup properly :(

File: pkg/server/test/list_objects.go [hunk 5]
```diff
@@ -51,6 +52,63 @@ func newListObjectsRequest(store, objectType, relation, user, modelID string, co
...
+		ctx = storage.ContextWithContextualTuples(ctx, request.ContextualTuples.GetTupleKeys())
+
+		resp, err := listObjectsQuery.Execute(ctx, request)
+		require.NoError(t, err)
+
+		objects := resp.GetObjects()
+		require.Len(t, objects, 2)
```
Comment (adriantam) [inline pkg/server/test/list_objects.go:108]: curious why do we need this instead of relying on the elements match below

File: pkg/server/test/list_objects.go (hunk 4 above)
Comment (jon-whit) [reply pkg/server/test/list_objects.go:101]: Yep! That's why we need the functional tests for ListObjects. Our test entrypoint should be `server.ListObjects` instead of `commands.ListObjectsQuery`.

File: pkg/server/test/list_objects.go (hunk 5 above)
Comment (jon-whit) [reply pkg/server/test/list_objects.go]: Good point - this is redundant. Removed 👍

**Review by adriantam — CHANGES_REQUESTED** [review]
Please add test to validate behaviour for duplicate contextual tuple for list object + check

**Review by adriantam — APPROVED** [review]

File: pkg/server/server.go [hunk 6]
```diff
@@ -153,11 +161,21 @@ func (s *Server) StreamedListObjects(req *openfgapb.StreamedListObjectsRequest,
...
 		return serverErrors.HandleError("", err)
 	}
 
 	typesys := typesystem.New(model)
 
+	for _, ctxTuple := range req.GetContextualTuples().GetTupleKeys() {
+		if err := validation.ValidateTuple(typesys, ctxTuple); err != nil {
+			return serverErrors.HandleTupleValidateError(err)
+		}
+	}
```
Comment (miparnisari) [inline pkg/server/server.go:174]: why isn't this also inside the non-streaming version of ListObjects?

Comment (miparnisari) [reply pkg/server/server.go:174]: also, this code is already inside `evaluate`... i'm confused

File: pkg/server/commands/list_objects.go [hunk 7]
```diff
@@ -58,21 +57,18 @@ func (q *ListObjectsQuery) handler(
...
 
-	model, err := q.Datastore.ReadAuthorizationModel(ctx, req.GetStoreId(), req.GetAuthorizationModelId())
-	if err != nil {
-		if errors.Is(err, storage.ErrNotFound) {
-			return serverErrors.AuthorizationModelNotFound(req.GetAuthorizationModelId())
-		}
-		return err
+	typesys, ok := typesystem.TypesystemFromContext(ctx)
+	if !ok {
+		panic("typesystem missing in context")
```
Comment (miparnisari) [inline pkg/server/commands/list_objects.go:62]: this should not be a panic since we can recover from this error by actually building the type system
